File: frontend/src/lib/permissions.ts
// Keep in lockstep with `backend/app/core/permissions.py`: the `Permission`
// union and `ROLE_PERMISSIONS` map below mirror the Python enum + role table
// by hand. When you add / rename / remove a permission on the backend, update
// both files in the same commit — otherwise the UI will hide or show controls
// that disagree with the API guard.
import { useMemo } from "react";
import { useAuth } from "../context/AuthContext";

export type UserRole = "owner" | "writer" | "stock_manager" | "factory_manager";

export type Permission =
  | "dashboard_view"
  | "reports_view"
  | "masters_read"
  | "masters_manage"
  | "book_settings_view"
  | "book_settings_edit"
  | "bills_manage"
  | "payments_manage"
  | "accounts_view"
  | "cashbook_manage"
  | "bank_accounts_manage"
  | "expense_categories_manage"
  | "fulfillment_write"
  | "fulfillment_view"
  | "job_work_manage"
  | "job_work_fulfillment_write"
  | "product_transfer_write"
  | "product_transfer_view"
  | "inventory_view"
  | "inventory_opening_stock"
  | "inventory_edit_qty"
  | "bag_change_write"
  | "bag_change_view"
  | "stock_disposal_write"
  | "stock_disposal_view"
  | "processing_manage"
  | "processing_view"
  | "void"
  | "users_manage"
  | "audit_view"
  | "notes_view";

const ROLE_PERMISSIONS: Record<UserRole, ReadonlySet<Permission>> = {
  owner: new Set<Permission>([
    "dashboard_view",
    "reports_view",
    "masters_read",
    "masters_manage",
    "book_settings_view",
    "book_settings_edit",
    "bills_manage",
    "payments_manage",
    "accounts_view",
    "cashbook_manage",
    "bank_accounts_manage",
    "expense_categories_manage",
    "fulfillment_write",
    "fulfillment_view",
    "job_work_manage",
    "job_work_fulfillment_write",
    "product_transfer_write",
    "product_transfer_view",
    "inventory_view",
    "inventory_opening_stock",
    "inventory_edit_qty",
    "bag_change_write",
    "bag_change_view",
    "stock_disposal_write",
    "stock_disposal_view",
    "processing_manage",
    "processing_view",
    "void",
    "users_manage",
    "audit_view",
    "notes_view",
  ]),
  writer: new Set([
    "dashboard_view",
    "reports_view",
    "masters_read",
    "inventory_view",
    "fulfillment_write",
    "fulfillment_view",
    "job_work_fulfillment_write",
    "product_transfer_write",
    "product_transfer_view",
    "notes_view",
  ]),
  stock_manager: new Set([
    "dashboard_view",
    "reports_view",
    "masters_read",
    "inventory_view",
    "bag_change_write",
    "bag_change_view",
    "stock_disposal_write",
    "stock_disposal_view",
    "notes_view",
  ]),
  factory_manager: new Set([
    "dashboard_view",
    "reports_view",
    "masters_read",
    "inventory_view",
    "book_settings_view",
    "processing_manage",
    "processing_view",
    "notes_view",
  ]),
};

export const ROLE_LABELS: Record<UserRole, string> = {
  owner: "Owner",
  writer: "Writer",
  stock_manager: "Stock manager",
  factory_manager: "Factory manager",
};

export function usePermissions() {
  const { user } = useAuth();
  const role = user?.role ?? null;

  return useMemo(() => {
    const permissions = role ? ROLE_PERMISSIONS[role] : new Set<Permission>();
    const can = (permission: Permission) => permissions.has(permission);
    return {
      role,
      hasRole: Boolean(role),
      can,
      isOwner: role === "owner",
      canVoid: can("void"),
      canManageUsers: can("users_manage"),
      canManageMasters: can("masters_manage"),
      canManageBills: can("bills_manage"),
      canManagePayments: can("payments_manage"),
      canViewAccounts: can("accounts_view"),
      canFulfillmentWrite: can("fulfillment_write"),
      canJobWorkFulfillment: can("job_work_fulfillment_write"),
      canJobWorkManage: can("job_work_manage"),
      canProductTransfer: can("product_transfer_write"),
      canBagChange: can("bag_change_write"),
      canStockDisposal: can("stock_disposal_write"),
      canProcessing: can("processing_manage"),
      canInventoryOpening: can("inventory_opening_stock"),
      canInventoryEditQty: can("inventory_edit_qty"),
      canViewBookSettings: can("book_settings_view"),
      canEditBookSettings: can("book_settings_edit"),
      canViewAuditLog: can("audit_view"),
    };
  }, [role]);
}

File: backend/app/core/permissions.py
"""Spec v15.0 — role-based access control."""

from enum import Enum
from typing import Callable

from fastapi import Depends, HTTPException

from app.core.auth import get_current_user
from app.models.entities import User, UserRole

ROLE_NOT_ASSIGNED_MSG = "No role assigned. Contact the owner to get access."
FORBIDDEN_MSG = "You do not have permission for this action."
OWNER_ONLY_VOID_MSG = "Only the owner role may void records."


class Permission(str, Enum):
    DASHBOARD_VIEW = "dashboard_view"
    REPORTS_VIEW = "reports_view"
    MASTERS_READ = "masters_read"
    MASTERS_MANAGE = "masters_manage"
    BOOK_SETTINGS_VIEW = "book_settings_view"
    BOOK_SETTINGS_EDIT = "book_settings_edit"
    BILLS_MANAGE = "bills_manage"
    PAYMENTS_MANAGE = "payments_manage"
    ACCOUNTS_VIEW = "accounts_view"
    CASHBOOK_MANAGE = "cashbook_manage"
    BANK_ACCOUNTS_MANAGE = "bank_accounts_manage"
    EXPENSE_CATEGORIES_MANAGE = "expense_categories_manage"
    FULFILLMENT_WRITE = "fulfillment_write"
    FULFILLMENT_VIEW = "fulfillment_view"
    JOB_WORK_MANAGE = "job_work_manage"
    JOB_WORK_FULFILLMENT_WRITE = "job_work_fulfillment_write"
    PRODUCT_TRANSFER_WRITE = "product_transfer_write"
    PRODUCT_TRANSFER_VIEW = "product_transfer_view"
    INVENTORY_VIEW = "inventory_view"
    INVENTORY_OPENING_STOCK = "inventory_opening_stock"
    INVENTORY_EDIT_QTY = "inventory_edit_qty"
    BAG_CHANGE_WRITE = "bag_change_write"
    BAG_CHANGE_VIEW = "bag_change_view"
    STOCK_DISPOSAL_WRITE = "stock_disposal_write"
    STOCK_DISPOSAL_VIEW = "stock_disposal_view"
    PROCESSING_MANAGE = "processing_manage"
    PROCESSING_VIEW = "processing_view"
    VOID = "void"
    USERS_MANAGE = "users_manage"
    AUDIT_VIEW = "audit_view"


ROLE_PERMISSIONS: dict[UserRole, frozenset[Permission]] = {
    UserRole.owner: frozenset(Permission),
    UserRole.writer: frozenset(
        {
            Permission.DASHBOARD_VIEW,
            Permission.REPORTS_VIEW,
            Permission.MASTERS_READ,
            Permission.INVENTORY_VIEW,
            Permission.FULFILLMENT_WRITE,
            Permission.FULFILLMENT_VIEW,
            Permission.JOB_WORK_FULFILLMENT_WRITE,
            Permission.PRODUCT_TRANSFER_WRITE,
            Permission.PRODUCT_TRANSFER_VIEW,
        }
    ),
    UserRole.stock_manager: frozenset(
        {
            Permission.DASHBOARD_VIEW,
            Permission.REPORTS_VIEW,
            Permission.MASTERS_READ,
            Permission.INVENTORY_VIEW,
            Permission.BAG_CHANGE_WRITE,
            Permission.BAG_CHANGE_VIEW,
            Permission.STOCK_DISPOSAL_WRITE,
            Permission.STOCK_DISPOSAL_VIEW,
        }
    ),
    UserRole.factory_manager: frozenset(
        {
            Permission.DASHBOARD_VIEW,
            Permission.REPORTS_VIEW,
            Permission.MASTERS_READ,
            Permission.INVENTORY_VIEW,
            Permission.BOOK_SETTINGS_VIEW,
            Permission.PROCESSING_MANAGE,
            Permission.PROCESSING_VIEW,
        }
    ),
}


def user_has_permission(user: User, permission: Permission) -> bool:
    if user.role is None:
        return False
    return permission in ROLE_PERMISSIONS.get(user.role, frozenset())


def require_assigned_role(user: User = Depends(get_current_user)) -> User:
    if user.role is None:
        raise HTTPException(status_code=403, detail=ROLE_NOT_ASSIGNED_MSG)
    return user


def require_permission(permission: Permission) -> Callable:
    def _dep(user: User = Depends(require_assigned_role)) -> User:
        if not user_has_permission(user, permission):
            raise HTTPException(status_code=403, detail=FORBIDDEN_MSG)
        return user

    return _dep


def require_any_permission(*permissions: Permission) -> Callable:
    def _dep(user: User = Depends(require_assigned_role)) -> User:
        if not any(user_has_permission(user, p) for p in permissions):
            raise HTTPException(status_code=403, detail=FORBIDDEN_MSG)
        return user

    return _dep


def require_roles(*roles: UserRole) -> Callable:
    allowed = frozenset(roles)

    def _dep(user: User = Depends(require_assigned_role)) -> User:
        if user.role not in allowed:
            raise HTTPException(status_code=403, detail=FORBIDDEN_MSG)
        return user

    return _dep


def require_owner_user(user: User = Depends(require_assigned_role)) -> User:
    if user.role != UserRole.owner:
        raise HTTPException(status_code=403, detail=OWNER_ONLY_VOID_MSG)
    return user


def require_void_user(user: User = Depends(require_owner_user)) -> User:
    if not user_has_permission(user, Permission.VOID):
        raise HTTPException(status_code=403, detail=OWNER_ONLY_VOID_MSG)
    return user
